lib/provider.js

```diff
@@ -115,15 +115,45 @@ const mergeConfigs = (providerConfig) => {
  * @param file
  * @returns string
  */
-const generateUploadFileName = (basePath, file) => {
-  const backupPath =
-    file.related && file.related.length > 0 && file.related[0].ref
-      ? `${file.related[0].ref}`
-      : `${file.hash}`;
-  const filePath = file.path ? `${file.path}/` : `${backupPath}/`;
-  const extension = file.ext.toLowerCase();
-  const fileName = slugify(path.basename(file.name + '_' + file.hash, file.ext)) + extension;
-  return `${basePath}${filePath}${fileName}`;
+const generateUploadFileName = (basePath, file, shortFilePaths) => {
+  if (!shortFilePaths) {
+    // Default version
+    const backupPath =
+      file.related && file.related.length > 0 && file.related[0].ref
+        ? `${file.related[0].ref}`
+        : `${file.hash}`;
+    const filePath = file.path ? `${file.path}/` : `${backupPath}/`;
+
+    // make sure that the extension is correct and the filename is valid:
+    // path.basename -> https://nodejs.org/docs/latest/api/path.html#path_path_basename_path_ext
+    // slugify -> https://www.npmjs.com/package/slugify
+    const fileName =
+      slugify(path.basename(file.name + '_' + file.hash, file.ext)) + file.ext.toLowerCase();
+
+    return `${basePath}${filePath}${fileName}`;
+  } else {
+    // New version with more intuitive subdirectories
+    // make sure that the extension is correct and the filename is valid:
+    // path.basename -> https://nodejs.org/docs/latest/api/path.html#path_path_basename_path_ext
+    // slugify -> https://www.npmjs.com/package/slugify
+    const fileName =
+      slugify(path.basename(file.hash, file.ext), { replacement: '-', remove: /[*+~.()'"!:@]/g }) +
+      file.ext.toLowerCase();
+
+    // Place all resized version of an image into one directory named after the original image
+    let fileDirectory = undefined;
+    ['thumbnail_', 'small_', 'medium_', 'large_'].forEach((prefix) => {
+      if (fileName.startsWith(prefix)) {
```

Comment:
I'm afraid with this cases, considering a user uploads a image with one of these prefixes.

Reply:
What if I replace all occurances of these prefixes (with the regex mentioned above)? The directory name would only be the same if a user would upload two images with the same hashes where the file name only differs in the prefix.

And even then inside the shared directory the filenames would still differ since I only crop the directory name like that.